From 79fa845d5213bf6acca096d8f3e2bb6c35ce8654 Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Fri, 12 Sep 2014 18:08:33 -0400
Subject: [powerpc] relocate fix relocate processing in LE mode

Message-id: <1410545655-205645-285-git-send-email-dzickus@redhat.com>
Patchwork-id: 94417
O-Subject: [RHEL7 PATCH 284/626] powerpc/relocate fix relocate processing in LE mode
Bugzilla: 1127366
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Steve Best <sbest@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: David Gibson <dgibson@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1127366

commit 3b830c824a8b15bf032eafe09c48a09d2441f16d
Author: Laurent Dufour <ldufour@linux.vnet.ibm.com>
Date:   Thu Jan 30 16:58:42 2014 +0100

    powerpc/relocate fix relocate processing in LE mode

    Relocation's code is not working in little endian mode because the r_info
    field, which is a 64 bits value, should be read from the right offset.

    The current code is optimized to read the r_info field as a 32 bits value
    starting at the middle of the double word (offset 12). When running in LE
    mode, the read value is not correct since only the MSB is read.

    This patch removes this optimization which consist to deal with a 32 bits
    value instead of a 64 bits one. This way it works in big and little endian
    mode.

    Signed-off-by: Laurent Dufour <ldufour@linux.vnet.ibm.com>
    Signed-off-by: Benjamin Herrenschmidt <benh@kernel.crashing.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/powerpc/kernel/reloc_64.S b/arch/powerpc/kernel/reloc_64.S
index b47a0e1..1482327 100644
--- a/arch/powerpc/kernel/reloc_64.S
+++ b/arch/powerpc/kernel/reloc_64.S
@@ -69,8 +69,8 @@ _GLOBAL(relocate)
   * R_PPC64_RELATIVE ones.
   */
  mtctr r8
-5: lwz r0,12(9) /* ELF64_R_TYPE(reloc->r_info) */
- cmpwi r0,R_PPC64_RELATIVE
+5: ld r0,8(9)  /* ELF64_R_TYPE(reloc->r_info) */
+ cmpdi r0,R_PPC64_RELATIVE
  bne 6f
  ld r6,0(r9) /* reloc->r_offset */
  ld r0,16(r9) /* reloc->r_addend */
-- 
1.7.1